### PR TITLE
HPA: expose the metrics "metric_computation_duration_seconds" and "metric_computation_total" from HPA controller

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature
/sig autoscaling
/sig instrumentation

#### What this PR does / why we need it:

expose the metrics "metric_computation_duration_seconds" and "metric_computation_total" from HPA controller.

- `metric_computation_duration_seconds`: The time(seconds) that the HPA controller takes to calculate one metric.
- `metric_computation_total`: Number of metric computations. 
These are must-to-have to meat the requirements from the production readiness review of [the container resource metrics](https://github.com/kubernetes/enhancements/issues/1610).

#### Which issue(s) this PR fixes:

Related https://github.com/kubernetes/kubernetes/issues/115639
Related https://github.com/kubernetes/enhancements/issues/1610

#### Special notes for your reviewer:

This branch is based on [sanposhiho:hpa-metric-server](https://github.com/kubernetes/kubernetes/pull/116010).
I know, in usual case, it's better to just wait for [sanposhiho:hpa-metric-server](https://github.com/kubernetes/kubernetes/pull/116010) to get merged, but I created this now because the code freeze is approaching and it's better to get reviews on it ASAP... 🏃‍♂️ 

Please review 3b1cde1e7c95d94c75da02a09291a59bb86e1b6f only. Other commits are from [sanposhiho:hpa-metric-server](https://github.com/kubernetes/kubernetes/pull/116010) branch.

#### Does this PR introduce a user-facing change?

```release-note
HPA controller exposes the following metrics from the kube-controller-manager.
- `metric_computation_duration_seconds`: Number of metric computations. 
- `metric_computation_total`: The time(seconds) that the HPA controller takes to calculate one metric.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/hold

It's ready to get reviews. 
But we should merge https://github.com/kubernetes/kubernetes/pull/116010 first.

See the `Special notes for your reviewer` in the issue description.

**Comment by sanposhiho** [comment]
/cc @logicalhan 
For the sig/instrumentation prospective reviews. 🙏

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

from sig instrumentation

**Comment by sanposhiho** [comment]
Just run go-fmt and re-push

File: pkg/controller/podautoscaler/monitor/metrics.go
Comment (gjtempleton) [inline pkg/controller/podautoscaler/monitor/metrics.go:29]: There's currently [efforts in other areas](https://github.com/kubernetes/kubernetes/pull/115813/files#diff-30acbb8d2e7a2aff26694443ae9353e09061f9a5aeb0f2703960371dda2e4e5cR65) to standardise the controller name scheme, with the HPA controller being referred to as `horizontal-pod-autoscaler-controller` there. Better to start with consistency with the proposed standard going forward unless there's a strong reason not to?

Comment (sanposhiho) [reply pkg/controller/podautoscaler/monitor/metrics.go:29]: 👍 
I'll rename here to `"horizontal_pod_autoscaler_controller"` in https://github.com/kubernetes/kubernetes/pull/116010. (replace `-` with `_` as no other metric named with `-`. ([ref](https://kubernetes.io/docs/reference/instrumentation/metrics/)))

**Comment by sanposhiho** [comment]
Just rebased.

File: pkg/controller/podautoscaler/monitor/metrics.go
Comment (gjtempleton) [inline pkg/controller/podautoscaler/monitor/metrics.go:53]:
```suggestion
			Help:           "Number of metric computations. The label 'action' should be either 'scale_down', 'scale_up', or 'none'. Also, the label 'error' should be either 'spec', 'internal', or 'none'. The label 'metric_type' corresponds to HPA.spec.metrics[*].type",
```

File: pkg/controller/podautoscaler/monitor/metrics.go
Comment (gjtempleton) [inline pkg/controller/podautoscaler/monitor/metrics.go:60]:
```suggestion
			Help:           "The time(seconds) that the HPA controller takes to calculate one metric. The label 'action' should be either 'scale_down', 'scale_up', or 'none'. The label 'error' should be either 'spec', 'internal', or 'none'. The label 'metric_type' corresponds to HPA.spec.metrics[*].type",
```

File: pkg/controller/podautoscaler/horizontal_test.go
Comment (gjtempleton) [inline pkg/controller/podautoscaler/horizontal_test.go:700]:
```suggestion
		t.Fatalf("the metric computation errors for %d types should be recorded, but actually only %d was recorded", len(tc.expectedReportedMetricComputationErrorLabels), len(m.metricComputationErrorLabels))
```

File: pkg/controller/podautoscaler/horizontal_test.go
Comment (gjtempleton) [inline pkg/controller/podautoscaler/horizontal_test.go:697]:
```suggestion
		t.Fatalf("the metric computation actions for %d types should be recorded, but actually only %d was recorded", len(tc.expectedReportedMetricComputationActionLabels), len(m.metricComputationActionLabels))
```

**Review by gjtempleton — COMMENTED** [review]
Minor nits.

**Comment by gjtempleton** [comment]
/lgtm
(I know this'll need to be re-applied once #116010 is merged.)

**Comment by sanposhiho** [comment]
Just rebased on the latest https://github.com/kubernetes/kubernetes/pull/116010

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Comment by mwielgus** [comment]
Please squash the commits, 10 is too much.

**Comment by sanposhiho** [comment]
@mwielgus Actually, this PR is only suppors to be https://github.com/kubernetes/kubernetes/commit/3b1cde1e7c95d94c75da02a09291a59bb86e1b6f as described in `Special notes for your reviewer`. I'll rebase this PR on the master once https://github.com/kubernetes/kubernetes/pull/116010 gets merged. 🙇

**Comment by sanposhiho** [comment]
Rebased on the latest https://github.com/kubernetes/kubernetes/pull/116010 anyway

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Comment by sanposhiho** [comment]
@mwielgus @gjtempleton Rebased.

**Comment by sanposhiho** [comment]
> Unfortunately the implementation PRs associated with this enhancement have not merged by code-freeze so this enhancement is getting removed from the release.

@marosset Sorry, I don't understand the reason.
This PR and https://github.com/kubernetes/kubernetes/pull/116046 are actually the implementation PR. And both of them are going to the merge queue now. Why do we need to `/hold` this?

**Comment by sanposhiho** [comment]
Also, see:

> All enhancements going into the release must be in a merge-ready state with approved and lgtm labels applied.
https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze

It should be OK if they have approved by the code freeze. (Not need to be merged by the code freeze)

**Comment by sanposhiho** [comment]
Discussion with the release team is on-going in:
https://kubernetes.slack.com/archives/CJH2GBF7Y/p1678843008739849

**Pull request merged**
